static/js/narrow.js

```diff
@@ -187,6 +189,19 @@ export function activate(raw_operators, opts) {
          or rerendering due to server-side changes.
     */
 
+    // Since narrow.activate is called directly from various
+    // places in our code without passing through hashchange,
+    // we need to check if the narrow is allowed for spectator here too.
+
+    if (
+        page_params.is_spectator &&
+        raw_operators.length &&
+        !hash_util.allowed_web_public_narrows.includes(raw_operators[0].operator)
```

Comment:
Hmm, why are we only looking at the first element of `raw_operators`?

Reply:
Okay, looking at all of them now. I thought I would do it after checking if this works but forgot.